networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/JvmLibP2PNetwork.java

```diff
@@ -158,4 +159,14 @@ public void stop() {
     host.stop();
     scheduler.shutdownNow();
   }
+
+  // Rest API methods
+
+  public String getPeerIDString() {
```

Comment:
You could make a single getPeerID method that returns the appropriate type (PeerID or String) depending on the needs of the calling code.  

```
  public <T> T getPeerId() {
      T result;
      Class<T> type = (Class<T>) ((ParameterizedType) getClass()
              .getGenericSuperclass())
              .getActualTypeArguments()[0];
      PeerId peerId = PeerId.fromPubKey(privKey.publicKey());
      if (type.isInstance(peerId)) {
          result = type.cast(peerId);
      }
      else {
          result = type.cast(peerId.toBase58());
      }
    return result;
  }
```

The following calls to the getPeerId method should then be valid:
```
String peerId = p2pNetwork.getPeerId();
PeerId peerId = p2pNetwork.getPeerId();
```

Reply:
really cool

Reply:
This seems quite overcomplicated.  Two methods would be much better (or just `getPeerId().toBase58()` in the calling site if it's not called too often).  Reflection is one of those tools that can be useful at times but definitely shouldn't be common.

Reply:
done

Reply:
I'll see myself out

Reply:
> This seems quite overcomplicated. Two methods would be much better (or just `getPeerId().toBase58()` in the calling site if it's not called too often). Reflection is one of those tools that can be useful at times but definitely shouldn't be common.

In languages where generics don't apply type erasure, this is a pretty normal thing to do bc you dont have to use reflection.  

I strongly recommend removing getPeerIdString....it is ugly and definitely not better.   Why not just call .toBase58() on it when needed.  It isn't much different than calling .toString()?

Reply:
I remember not being able to call getPeerId() from the RestAPI module because it didn't know the libp2p's data structure. So I couldn't do .toBase58() there. I'm not sure if that can be easily mitigated by importing the libp2p module in the caller module tho.

Reply:
ya... u could certainly import it, but maybe it's not worth it.  fair